File: submitted code/20.valid-parentheses.java
//
// @lc app=leetcode.cn id=20 lang=java
//
// [20] valid-parentheses
//
class Solution {
    public boolean isValid(String s) {
        int n = s.length();
        if (n % 2 == 1)
            return false;
        Deque<Character> deque = new ArrayDeque<>();
        for (int i = 0; i < n; i++) {
            if (s.charAt(i) == '(') {
                deque.addFirst('(');
            }
            else if (s.charAt(i) == ')') {
                if(deque.isEmpty() || deque.peek() != '(')
                    return false;
                deque.poll();
            }
            else if (s.charAt(i) == '[') {
                deque.addFirst('[');
            }
            else if (s.charAt(i) == ']'){
                if(deque.isEmpty() || deque.peek() != '[')
                    return false;
                deque.poll();
            }
            else if (s.charAt(i) == '{') {
                deque.addFirst('{');
            }
            else {
                if(deque.isEmpty() || deque.peek() != '{')
                    return false;
                deque.poll();
            }
        }
        return deque.isEmpty();
    }
}
// @lc code=end

File: submitted_code/20.valid-parentheses.c
//
// @lc app=leetcode.cn id=20 lang=c
//
// [20] valid-parentheses
//
bool isValid(char * s){

    if (s==NULL || strlen(s)<=0) return true;
    int top = 0;
    char *stack = (char*)malloc(strlen(s)+1);
    for (int i = 0; i<strlen(s); i++){
        if(s[i]=='(' || s[i]=='{' || s[i]=='['){
            stack[top++] = s[i];
        }else{
            if(--top < 0) return false;
            if(s[i]==')' && stack[top] != '(') return false;
            if(s[i]=='}' && stack[top] != '{') return false;
            if(s[i]==']' && stack[top] != '[') return false;
        }
    }
    if (top > 0) return false;
    return true;
}
// @lc code=end